src/main/kotlin/com/haroldadmin/cnradapter/NetworkResponse.kt

```diff
@@ -1,61 +1,105 @@
 package com.haroldadmin.cnradapter
 
 import okhttp3.Headers
+import retrofit2.Response
 import java.io.IOException
 
-sealed class NetworkResponse<out T : Any, out U : Any> {
+/**
+ * Represents the result of a network call made using Retrofit.
+ *
+ * A NetworkResponse can be either in a success state or an error state,
+ * depending on the result of the network call.
+ *
+ * If the network call was successful, the [NetworkResponse] is [NetworkResponse.Success].
+ * However, if the server responded with a 2xx status code WITHOUT a body, the success body type ([S])
+ * must be set to [Unit] OR the response's status code must be 204.
+ * If either of these conditions are true, then the response will contain [Unit] as the body.
+ * Otherwise, the [NetworkResponse] will be [NetworkResponse.Error.ServerError].
+ *
+ * If the network call failed due to:
+ * - Non-2xx response from the server, the [NetworkResponse] is [NetworkResponse.Error.ServerError]
+ * - Internet connectivity problems, the [NetworkResponse] is [NetworkResponse.Error.NetworkError]
+ * - Any other problems (e.g. Serialization errors), the [NetworkResponse] is [NetworkResponse.Error.UnknownError].
+ */
+public sealed interface NetworkResponse<S, E> {
     /**
-     * A request that resulted in a response with a 2xx status code that has a body.
-     */
-    data class Success<T : Any>(
-        val body: T,
-        val headers: Headers? = null,
-        val code: Int
-    ) : NetworkResponse<T, Nothing>()
-
-    /**
-     * Describe an error without a specific type.
-     * Makes it easier to deal with the case where you just want to know that an error occurred,
-     * without knowing the type
-     *
-     * @example
-     * val response = someNetworkAction()
+     * The result of a successful network request.
      *
-     * when (response) {
-     *    is NetworkResponse.Success -> // Action Succeeded do something with body
+     * If you expect your server response to not contain a body, set the success body type ([S]) to [Unit],
```

Comment:
I think there should be a `NetworkResponse.Complete` and there can be a discussion on if we should include 
`public val response: Response<*>` or not

we can also add a typealias of CompletableResponse

`typealias CompletableResponse<E> = NetworkResponse<Unit, E>`

Reply:
I've included the raw `Response` as an escape hatch when you want more control over the response than what this library provides. I see no harm in adding it, but a loss of control for consumers in removing it. Unless there's a strong argument against not including it, it's better to leave it in there IMO.

Great idea on the `CompletableResponse`, I'll add it as a nice utility.